### PR TITLE
FLI-1258: pass org and teams to metadata

### LINKED ISSUE #3435: [FLI-1258] Allow passing GitHub 'claims'/metadata to Authz
### Problem

Re: https://community.flipt.io/t/restrict-access-to-feature-flags-per-teams/41/2?u=mark

We currently map OIDC authn `claims` to a claims field in the metadata that is passed to our authz engine (ie: https://docs.flipt.io/guides/operation/authorization/rbac-with-keycloak), however we don't provide similar functionality for GitHub auth.

The user in the above linked Community post would like to create authz policies based on the authenticated user's GitHub team membership

### Ideal Solution

Provide a way to pass an authenticated GitHub user's data to our authz engine. It would be nice if it were similar to how we do it in OIDC using something like `claims`, although I realize GitHub user response doesn't have a field called `claims` so maybe we need to name it something else?

https://docs.github.com/en/rest/users/users?apiVersion=2022-11-28#get-the-authenticated-user

It also doesn't return the organization/teams of a user without doing another REST API call like we do to support Github allowed [orgs](https://docs.flipt.io/configuration/authentication#allowed-organizations)/[teams](https://docs.flipt.io/configuration/authentication#allowed-teams). But since we are already doing this at the authn stage, we could add this data to the metadata passed to authz.

Maybe we use something more generically named than `claims` for GitHub / other non-OIDC authn methods?

### Search

- [X] I searched for other open and closed issues before opening this

### Additional Context

_No response_

<sub>[FLI-1258](https://linear.app/flipt/issue/FLI-1258/allow-passing-github-claimsmetadata-to-authz)</sub>

### PR DESCRIPTION
**Pull request by devumesh**
**Approach:**
1. Get orgs and teams of the authenticated user by default. I went with this approach because there might be a case where authentication is not needed based on the orgs or teams but authorization might be needed to control RBAC.
2. Validate authentication on orgs `if orgs > 0`.
3. Validate team `if allowed_teams > 0`.
4. Convert orgs to list (stringified) and store in `io.flipt.auth.github.orgs`.
5. Convert teams to map `{org1: [list of teams],org2: [list of teams]}` and store in `io.flipt.auth.github.teams`.

closes #3435

### CODE REVIEW

**Review by erka — CHANGES_REQUESTED** [review]
Thank you for the contribution @devumesh!

There are few things which we should address in this PR.

Original code hides some aspects of  Github login integration and tests does not cover it.  There should be a token with `read:org` in scope to fetch user's orgs and teams. Originally the check happens on the configuration level https://github.com/flipt-io/flipt/blob/main/internal/config/authentication.go#L600-L602.  Without that  `getUserOrgs` and `getUserTeamsByOrg` api requests will fail and auth flow will be broken for those who don't have that scope in their Flipt configuration. 

Also I would like to discuss another thing. I think we should not store all user's orgs and teams, if there is a configuration requirement for specific organizations and/or teams for privacy reasons. wdyt? cc: @markphelps

File: internal/server/authn/method/github/server.go [hunk 1]
```diff
@@ -43,6 +43,8 @@ const (
 	storageMetadataGithubPicture           = "io.flipt.auth.github.picture"
 	storageMetadataGithubSub               = "io.flipt.auth.github.sub"
 	storageMetadataGitHubPreferredUsername = "io.flipt.auth.github.preferred_username"
+	storageMetadataGitHubOrgs              = "io.flipt.auth.github.orgs"
```
Comment (markphelps) [inline internal/server/authn/method/github/server.go:46]:
```suggestion
	storageMetadataGitHubOrgs              = "io.flipt.auth.github.organizations"
```

File: internal/server/authn/method/github/server.go [hunk 2]
```diff
@@ -165,26 +167,24 @@ func (s *Server) Callback(ctx context.Context, r *auth.CallbackRequest) (*auth.C
 	set(method.StorageMetadataName, githubUserResponse.Name)
 	set(method.StorageMetadataPicture, githubUserResponse.AvatarURL)
 
-	if len(s.config.Methods.Github.Method.AllowedOrganizations) != 0 {
-		userOrgs, err := getUserOrgs(ctx, token, apiURL)
-		if err != nil {
-			return nil, err
-		}
+	userOrgs, err := getUserOrgs(ctx, token, apiURL)
```
Comment (markphelps) [inline internal/server/authn/method/github/server.go:170]: note for another PR: we should consider caching these results in memory for a period of time as now we're doing these REST calls on every request if an org is using GitHub authn and the orgs/team membership do not change nearly that often

**Review by markphelps — CHANGES_REQUESTED** [review]
thanks @devumesh ! I agree with @erka that we could use a bit more testing.

@erka im not too concerned about storing all the orgs/team memberships in the metadata tbh. how are you thinking we would limit what is stored in metadata? another configuration option?

**Comment by erka** [comment]
> @erka im not too concerned about storing all the orgs/team memberships in the metadata tbh. how are you thinking we would limit what is stored in metadata? another configuration option?

Let say, that Flipt configuration has `allowed_organizations` which set to `[a, b]`. And user has orgs `[a, d, f, g, e, t]`. Is there any usecase when those `d, f, g, e, t` could be useful? I think that only allowed_organizations should be added to metadata in this case.

**Comment by markphelps** [comment]
> > @erka im not too concerned about storing all the orgs/team memberships in the metadata tbh. how are you thinking we would limit what is stored in metadata? another configuration option?
> 
> 
> 
> Let say, that Flipt configuration has `allowed_organizations` which set to `[a, b]`. And user has orgs `[a, d, f, g, e, t]`. Is there any usecase when those `d, f, g, e, t` could be useful? I think that only allowed_organizations should be added to metadata in this case.

Ah that's a good point. I was thinking you were suggesting another config property but I agree with you we should only track the ones configured.

**Comment by devumesh** [comment]
> Let say, that Flipt configuration has `allowed_organizations` which set to `[a, b]`. And user has orgs `[a, d, f, g, e, t]`. Is there any usecase when those `d, f, g, e, t` could be useful? I think that only allowed_organizations should be added to metadata in this case.

I have a doubt. If the user doesn't configure `allowed_organizations`, should we need to pass `[a, d, f, g, e, t]` orgs for authz or authn is needed for authz with github?

**Comment by erka** [comment]
> Hi @erka, correct me If I'm understanding wrong. `read:org` in scope should me mandated when using Github?

Maybe if your Flipt instance is in internet. But I don't think we can't change it, at least in v1.

It started with `user:email` scope. One day someone asked for organization filter because of access control concern. I believe they had the Flipt instance in public network. Later someone else asked for teams support.  For these orgs and team filters `read:org` is needed.  The [documentation](https://docs.flipt.io/configuration/authentication#github)  looks like a history in some sense. 

It would be great to not break auth flow for those first adopters of login with Github.

**Comment by erka** [comment]
> I have a doubt. If the user doesn't configure `allowed_organizations`, should we need to pass `[a, d, f, g, e, t]` orgs for authz or authn is needed for authz with github?

Yeah, it's tricky... If there is `read:org` in configuration scopes , let's read them and add to metadata, if not - we can't fetch them and they should be empty.  wdyt?

**Comment by devumesh** [comment]
> Yeah, it's tricky... If there is `read:org` in configuration scopes , let's read them and add to metadata, if not - we can't fetch them and they should be empty. wdyt?

Yeah I guess it should work. If the allowed org is not provided are you suggesting to send all the retrieved orgs for authz?
or
Authz for github with org and teams valid only with allowed orgs for authn?

Can you please clarify this.

File: internal/server/authn/method/github/server.go (hunk 1 above)
Comment (devumesh) [reply internal/server/authn/method/github/server.go]: Hi @markphelps , fixed this

**Comment by erka** [comment]
@devumesh Let me explain my vision with examples so you could validate them.

#### 1. no read:org in scopes 
No way to get user's organizations and teams, so metadata for organizations and teams should be empty

#### 2. read:org in scopes

##### 2.1. no organizations restrictions
Fetch user's orgs and teams and add them all to metadata. For example, user has `a, b, e` orgs, so metadata will have `a, b, e` too. 

##### 2.2. with organization restrictions
Fetch user's orgs and teams, filter them and include only required organizations which user has. As example, user has `a, b, c, d` organizations and restriction contains `a, b, e`. Only `a, b` orgs and teams of `a, b` should be included in metadata.

##### 2.3. with organization and team restrictions
Same as 2.2 but filter user's teams as well. 

WDYT?

**Comment by devumesh** [comment]
> Fetch user's orgs and teams, filter them and include only required organizations which user has. As example, user has `a, b, c, d` organizations and restriction contains `a, b, e`. Only `a, b` orgs and teams of `a, b` should be included in metadata.

Hey @erka , thanks. I have made changes as you mentioned. But I have made a mistake where instead of passing orgs `a, b` I'm passing `a, b, e` (passing all the orgs mentioned in the org restriction). Same mistake I made for teams also. I will fix that.

**Comment by markphelps** [comment]
@all-contributors please add @devumesh for code

**Review by markphelps — APPROVED** [review]
Thank you!

**Comment by devumesh** [comment]
Hi @markphelps, thank you. committed your suggestion. Please review.

**Pull request merged**